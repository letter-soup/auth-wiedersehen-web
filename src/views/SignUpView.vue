<script setup lang="ts">
import { Carousel, CarouselContent, CarouselItem } from '@/components/ui/carousel'
import { Button } from '@/components/ui/button'
import { Input } from '@/components/ui/input'
import { Label } from '@/components/ui/label'
import { Toggle } from '@/components/ui/toggle'
import { ArrowLeft } from 'lucide-vue-next'
import { useRouter } from 'vue-router';
import { computed, type ComputedRef, ref, type Ref } from 'vue'

const { currentRoute, back, push } = useRouter();
type SignUpStep = 'email' | 'password' | 'tos'

const currentStep: Ref<SignUpStep> = ref('email')

const canGoBack: ComputedRef<boolean> = computed(() => currentStep.value !== 'email')

function validateEmail(scrollNext: () => void, next: SignUpStep): void {
  currentStep.value = next
  scrollNext()
}

</script>

<template>
  <div class="w-full lg:grid lg:min-h-[600px] lg:grid-cols-2 xl:min-h-[800px]">
    <div class="flex items-center justify-center py-12">
      <div class="mx-auto grid w-[350px] gap-6">
        <div class="grid gap-2 text-center">
          <h1 class="text-3xl font-bold">
            {{ $t('sign-up:header') }}
          </h1>
          <p class="text-balance text-muted-foreground">
            {{ $t('sign-up:description') }}
          </p>
        </div>
        <div class="grid gap-4">
          <form>
            <Carousel
              v-slot="{ scrollNext, scrollPrev }"
              class="relative w-full max-w-xs"
              :opts="{ watchDrag: false, duration: 0 }"
            >
              <Toggle v-if="canGoBack" @click="scrollPrev">
                <ArrowLeft class="h-4 w-4" />
              </Toggle>
              <CarouselContent>
                <CarouselItem>
                  <div class="grid gap-2">
                    <Label for="email">{{ $t('sign-up:email') }}</Label>
                    <Input id="email" type="email" placeholder="m@example.com" required />
                    <Button @click="validateEmail(scrollNext)"> Next </Button>
                  </div>
                </CarouselItem>
                <CarouselItem>
                  <div class="grid gap-2">
                    <Label for="password">{{ $t('sign-up:password') }}</Label>
                    <Input id="password" type="password" required />
                  </div>
                </CarouselItem>
              </CarouselContent>
            </Carousel>
          </form>
        </div>
        <div class="mt-4 text-center text-sm">
          {{ $t('sign-up:has-account') }}
          <a href="/sign-up" class="underline">
            {{ $t('sign-up:sign-in') }}
          </a>
        </div>
      </div>
    </div>
    <div class="hidden bg-muted lg:block">
      <img
        src="https://cdn.pixabay.com/photo/2024/06/01/12/28/monday-8802107_1280.jpg"
        alt="image"
        class="h-dvh w-full object-cover dark:brightness-[0.2] dark:grayscale"
      />
    </div>
  </div>
</template>
