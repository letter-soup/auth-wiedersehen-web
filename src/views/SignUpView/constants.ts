import * as z from 'zod'
import type { SignUpStep } from '@/views/SignUpView/types'

const lowerCaseSymbols = /[a-z]+/
const upperCaseSymbols = /[A-Z]+/
const numericSymbols = /[0-9]+/
const specialSymbols = /[!@#$%^&*()\-_+.]+/

const passwordErrorMessage = 'Password must have at least 8 characters, lowercase and uppercase letters, numbers and special symbols'

export const formSchema = [
  z.object({
    email: z.string().email(),
  }),
  z
    .object({
      password: z.string()
        .min(8, passwordErrorMessage)
        .regex(lowerCaseSymbols, passwordErrorMessage)
        .regex(upperCaseSymbols, passwordErrorMessage)
        .regex(numericSymbols, passwordErrorMessage)
        .regex(specialSymbols, passwordErrorMessage),
      confirmPassword: z.string()
    })
    .refine(
      (values) => {
        return values.password === values.confirmPassword
      },
      {
        message: 'Passwords must match!',
        path: ['confirmPassword']
      }
    )
]

export const signUpSteps: SignUpStep[] = [
  {
    step: 1,
    type: 'email',
    description: 'sign-up:step-email:description'
  },
  {
    step: 2,
    type: 'password',
    description: 'sign-up:step-password:description'
  },
]
