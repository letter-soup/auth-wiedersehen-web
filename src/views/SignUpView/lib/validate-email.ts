import type { TFormValidationCallback } from '@/lib/types'

export async function validateEmail(
  values: Record<string, string>,
  validate: TFormValidationCallback,
  nextStep: () => void,
  setFieldError: (field: string, error: string) => void,
  t: (key: string) => string,
) {
  const validationResult = await validate()
  if (!validationResult.valid) {
    return
  }

  if (validationResult.valid) {
    nextStep()
  }
}
