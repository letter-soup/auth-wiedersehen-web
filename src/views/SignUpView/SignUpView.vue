<script setup lang="ts">
import { toTypedSchema } from '@vee-validate/zod'
import { computed, type ComputedRef, h, type Ref, ref, watch } from 'vue'
import { useRouter } from 'vue-router'
import { Stepper, StepperItem, StepperTrigger } from '@/components/ui/stepper'
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from '@/components/ui/form'
import { Input } from '@/components/ui/input'
import { Button } from '@/components/ui/button'
import { toast } from '@/components/ui/toast'
import { Separator } from '@/components/ui/separator'
import { SIGN_UP_STEPS } from '@/views/SignUpView/lib/constants'
import { useI18n } from 'vue-i18n'
import type { TFormValidationCallback } from '@/lib/types'
import { createFormSchema } from '@/views/SignUpView/lib/form-schema'
import { validateEmail } from '@/views/SignUpView/lib/validate-email'
import { useRouterGuard } from '@/composable/useRouterGuard'

const { t } = useI18n()
const router = useRouter()

const initialStep: number = 1
const stepIndex: Ref<number> = ref(initialStep)
const stepDescription: ComputedRef<string> = computed(
  () => SIGN_UP_STEPS[stepIndex.value - 1].description,
)
const formSchema: ComputedRef = computed(() => createFormSchema(t))

useRouterGuard(() =>
  router.beforeEach((to, from, next) => {
    if (to.name !== from.name) {
      return next()
    }

    const toStep = parseInt(to.query.step as string) || initialStep
    if (toStep < stepIndex.value) {
      stepIndex.value = toStep
    }
    return next()
  }),
)

watch(
  () => stepIndex.value,
  (value) => {
    router.push({
      ...router.currentRoute.value,
      query: { step: value },
    })
  },
)

const signUpAcknowledgementMap: ComputedRef<Record<string, string>> = computed(() => ({
  termsOfService: `<a target="_blank" href="/terms" class="underline underline-offset-4 hover:text-primary">${t('sign-up:terms-of-service')}</a>`,
  privacyPolicy: `<a target="_blank" href="/privacy" class="underline underline-offset-4 hover:text-primary">${t('sign-up:privacy-policy')}</a>`,
}))

async function onSubmit(
  e: Event,
  validate: TFormValidationCallback,
  values: Record<string, string>,
) {
  const validationResult = await validate()

  if (stepIndex.value === SIGN_UP_STEPS.length && validationResult.valid) {
    toast({
      title: 'You submitted the following values:',
      description: h(
        'pre',
        { class: 'mt-2 w-[340px] rounded-md bg-slate-950 p-4' },
        h('code', { class: 'text-white' }, JSON.stringify(values, null, 2)),
      ),
    })
  }
}
</script>

<template>
  <div class="w-full h-full flex items-center justify-center py-12">
    <div class="mx-auto grid w-[350px] gap-6">
      <div class="grid gap-2 text-center">
        <h1 class="text-3xl font-bold">
          {{ $t('sign-up:header') }}
        </h1>
        <p class="text-balance text-muted-foreground">
          {{ $t(stepDescription) }}
        </p>
      </div>
      <Form
        v-slot="{ values, validate, setFieldError }"
        as=""
        keep-values
        :validation-schema="toTypedSchema(formSchema[stepIndex - 1])"
      >
        <Stepper v-slot="{ nextStep }" v-model="stepIndex" class="block w-full">
          <form @submit.prevent="(e) => onSubmit(e, validate, values)">
            <div class="flex w-full flex-start gap-2">
              <StepperItem
                v-for="step in SIGN_UP_STEPS"
                :key="step.step"
                class="hidden"
                :step="step.step"
              >
                <StepperTrigger as-child>
                  <div />
                </StepperTrigger>
              </StepperItem>
            </div>

            <div
              class="flex flex-col gap-4"
              v-if="stepIndex === 1"
              data-testid="sign-up-step-email"
            >
              <FormField v-slot="{ componentField }" name="email">
                <FormItem>
                  <FormLabel>{{ $t('sign-up:email') }}</FormLabel>
                  <FormControl>
                    <Input
                      id="email"
                      type="email"
                      v-bind="componentField"
                      @keydown.enter.prevent="
                        validateEmail(values, validate, nextStep, setFieldError, $t)
                      "
                    />
                  </FormControl>
                  <FormMessage />
                </FormItem>
              </FormField>

              <Button
                id="cta"
                class="w-full"
                @click="validateEmail(values, validate, nextStep, setFieldError, $t)"
                type="button"
              >
                {{ $t('sign-up:step:cta') }}
              </Button>
            </div>

            <div
              class="flex flex-col gap-4"
              v-if="stepIndex === 2"
              data-testid="sign-up-step-password"
            >
              <FormField v-slot="{ componentField }" name="password">
                <FormItem>
                  <FormLabel>{{ $t('sign-up:password') }}</FormLabel>
                  <FormControl>
                    <Input id="password" type="password" v-bind="componentField" />
                  </FormControl>
                  <FormMessage />
                </FormItem>
              </FormField>

              <FormField v-slot="{ componentField }" name="confirmPassword">
                <FormItem>
                  <FormLabel>{{ $t('sign-up:confirm-password') }}</FormLabel>
                  <FormControl>
                    <Input id="confirm-password" type="password" v-bind="componentField" />
                  </FormControl>
                  <FormMessage />
                </FormItem>
              </FormField>

              <Button type="submit" class="w-full">
                {{ $t('sign-up:cta') }}
              </Button>

              <p
                class="px-8 text-center text-sm text-muted-foreground"
                v-html="$t('sign-up:cta-acknowledgement', signUpAcknowledgementMap)"
              />
            </div>
          </form>
        </Stepper>
      </Form>

      <Separator class="mt-2" :label="$t('sign-up:has-account')" orientation="horizontal" />
      <Button variant="link">
        <RouterLink to="/sign-in">
          {{ $t('sign-up:sign-in') }}
        </RouterLink>
      </Button>
    </div>
  </div>
</template>
