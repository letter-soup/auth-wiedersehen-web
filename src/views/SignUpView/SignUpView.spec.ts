import '@/views/SignUpView/lib/__mock__/validate-email.mock'
import { beforeEach, describe, expect, it } from 'vitest'
import { flushPromises } from '@vue/test-utils'
import { mount } from '@/lib/tests/mount'
import SignUpView from '@/views/SignUpView/SignUpView.vue'

describe('SignUpView (snapshot)', () => {
  it('should render email step by default', async () => {
    const wrapper = mount(SignUpView)
    expect(wrapper).toBeTruthy()

    expect(wrapper.find('[data-testid=sign-up-step-email]').exists()).toBeTruthy()
    expect(wrapper.find('[data-testid=sign-up-step-password]').exists()).toBeFalsy()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('given valid email, should render password step on cta click', async () => {
    const wrapper = mount(SignUpView)
    expect(wrapper).toBeTruthy()

    await wrapper.find('[id=email]').setValue('valid@email.com')
    await wrapper.find('[id=cta]').trigger('click')
    await flushPromises()

    expect(wrapper.find('[data-testid=sign-up-step-email]').exists()).toBeFalsy()
    expect(wrapper.find('[data-testid=sign-up-step-password]').exists()).toBeTruthy()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
