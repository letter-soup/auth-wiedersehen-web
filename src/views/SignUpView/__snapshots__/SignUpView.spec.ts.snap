// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignUpView (snapshot) > given valid email, should render password step on cta click 1`] = `
"<div class="w-full h-full flex items-center justify-center py-12">
  <div class="mx-auto grid w-[350px] gap-6">
    <div class="grid gap-2 text-center">
      <h1 class="text-3xl font-bold">sign-up:header</h1>
      <p class="text-balance text-muted-foreground">sign-up:step-password:description</p>
    </div>
    <div role="group" aria-label="progress" data-linear="" data-orientation="horizontal" class="gap-2 block w-full">
      <form>
        <div class="flex w-full flex-start gap-2">
          <div data-state="completed" data-orientation="horizontal" class="items-center gap-2 group data-[disabled]:pointer-events-none hidden">
            <div type="button" data-state="completed" data-orientation="horizontal" tabindex="0" aria-describedby="radix-vue-stepper-item-description-v-1" aria-labelledby="radix-vue-stepper-item-title-v-0" class="p-2 flex flex-col items-center text-center gap-2 rounded-md"></div>
          </div>
          <div data-state="active" data-orientation="horizontal" class="items-center gap-2 group data-[disabled]:pointer-events-none hidden" aria-current="true">
            <div type="button" data-state="active" data-orientation="horizontal" tabindex="0" aria-describedby="radix-vue-stepper-item-description-v-3" aria-labelledby="radix-vue-stepper-item-title-v-2" class="p-2 flex flex-col items-center text-center gap-2 rounded-md"></div>
          </div>
        </div>
        <!--v-if-->
        <div class="flex flex-col gap-4" data-testid="sign-up-step-password">
          <div class="space-y-2"><label for="radix-v-5-form-item" class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70">sign-up:password</label><input class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" id="password" type="password" name="password" aria-describedby="radix-v-5-form-item-description" aria-invalid="false">
            <!---->
          </div>
          <div class="space-y-2"><label for="radix-v-6-form-item" class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70">sign-up:confirm-password</label><input class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" id="confirm-password" type="password" name="confirmPassword" aria-describedby="radix-v-6-form-item-description" aria-invalid="false">
            <!---->
          </div><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full" type="submit">sign-up:cta</button>
          <p class="px-8 text-center text-sm text-muted-foreground">sign-up:cta-acknowledgement</p>
        </div>
      </form>
      <div aria-live="polite" aria-atomic="true" role="status" style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px;"> Step 2 of 2</div>
    </div>
    <div data-orientation="horizontal" role="separator" label="sign-up:has-account" class="shrink-0 bg-border relative h-px w-full mt-2"><span class="text-xs text-muted-foreground bg-background absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex justify-center items-center h-[1px] py-1 px-2">sign-up:has-account</span></div><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 text-primary underline-offset-4 hover:underline h-10 px-4 py-2"><a href="/sign-in">sign-up:sign-in</a></button>
  </div>
</div>"
`;

exports[`SignUpView (snapshot) > should render email step by default 1`] = `
"<div class="w-full h-full flex items-center justify-center py-12">
  <div class="mx-auto grid w-[350px] gap-6">
    <div class="grid gap-2 text-center">
      <h1 class="text-3xl font-bold">sign-up:header</h1>
      <p class="text-balance text-muted-foreground">sign-up:step-email:description</p>
    </div>
    <div role="group" aria-label="progress" data-linear="" data-orientation="horizontal" class="gap-2 block w-full">
      <form>
        <div class="flex w-full flex-start gap-2">
          <div aria-current="true" data-state="active" data-orientation="horizontal" class="items-center gap-2 group data-[disabled]:pointer-events-none hidden">
            <div type="button" data-state="active" data-orientation="horizontal" tabindex="0" aria-describedby="radix-vue-stepper-item-description-v-1" aria-labelledby="radix-vue-stepper-item-title-v-0" class="p-2 flex flex-col items-center text-center gap-2 rounded-md"></div>
          </div>
          <div data-state="inactive" data-orientation="horizontal" class="items-center gap-2 group data-[disabled]:pointer-events-none hidden">
            <div type="button" data-state="inactive" data-orientation="horizontal" tabindex="0" aria-describedby="radix-vue-stepper-item-description-v-3" aria-labelledby="radix-vue-stepper-item-title-v-2" class="p-2 flex flex-col items-center text-center gap-2 rounded-md"></div>
          </div>
        </div>
        <div class="flex flex-col gap-4" data-testid="sign-up-step-email">
          <div class="space-y-2"><label for="radix-v-4-form-item" class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70">sign-up:email</label><input class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" id="email" type="email" name="email" aria-describedby="radix-v-4-form-item-description" aria-invalid="false">
            <!---->
          </div><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full" id="cta" type="button">sign-up:step:cta</button>
        </div>
        <!--v-if-->
      </form>
      <div aria-live="polite" aria-atomic="true" role="status" style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px;"> Step 1 of 0</div>
    </div>
    <div data-orientation="horizontal" role="separator" label="sign-up:has-account" class="shrink-0 bg-border relative h-px w-full mt-2"><span class="text-xs text-muted-foreground bg-background absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex justify-center items-center h-[1px] py-1 px-2">sign-up:has-account</span></div><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 text-primary underline-offset-4 hover:underline h-10 px-4 py-2"><a href="/sign-in">sign-up:sign-in</a></button>
  </div>
</div>"
`;
